rfcs/2022-02-11-responsive-and-adaptive-layout-strategy.md

```diff
@@ -0,0 +1,414 @@
+
+**1. Components that can handle Responsiveness Internally**
+
+For this to work, on the consumer side nothing has to be done explicitly, as far as the consumers have wrapped their app inside `<BladeProvider><ConsumerApp></BladeProvider>`, the components which will handle responsiveness internally will automatically listen to the screen size changes and scale automatically based on the available space
+
+**2. Components that can give some flexibility to handle Responsiveness to consumers**
+
+Consider, if we want to implement a layout similar to the below mockup
+<img alt="Adaptive Layout with flex direction" src="./images/responsive-adaptive-layout/adaptive-layout-with-flex-direction.png">
+
+<br/>
+
+Here's how the consumer will create a layout like this in code(_Note: the actual API might differ_)
+
+```jsx
+import { Stack, Card, Title } from '@razorpay/blade/components';
+
+const Dashboard = () => (
+  <Stack direction={{ s: 'column', l: 'row' }}>
+    <Card>
+      <Title>This is Card Title</Title>
+      <Content>This is Card Content</Content>
+    </Card>
+    <Card>
+      <Title>This is Card Title</Title>
+      <Content>This is Card Content</Content>
+    </Card>
+    <Card>
+      <Title>This is Card Title</Title>
+      <Content>This is Card Content</Content>
+    </Card>
+  </Stack>
+);
+
+export default Dashboard;
+```
+
+**3. Components for Adaptive layout(rendering components conditionally)**
+
+- Consider, if we want to implement a layout similar to the below mockup where we have 2 different search inputs for different screens of the same web app.
+  <img alt="Responsive Modal" src="./images/responsive-adaptive-layout/adaptive-search-field.png">
+
+  <br/>
+
+  Here's how the consumer will create a layout like this in code(_Note: the actual API might differ_).
+
+  ```jsx
+  import { LargeSearchInput, SmallSearchInput } from '@razorpay/blade/components';
+
+  const SearchScreen = () => (isMobile ? <SmallSearchInput /> : <LargeSearchInput />);
+
+  export default SearchScreen;
+  ```
+
+  Now, there's an issue with this approach. If you look we are importing both the components(LargeSearchInput, SmallSearchInput) for both the screens(Desktop/Mobile) which will unnecessary increase the bundlesize and hamper the performance. So to solve this we shall use [`React.Lazy`](https://reactjs.org/docs/code-splitting.html#reactlazy) if you're doing client side rendering or [`@loadable/component`](https://loadable-components.com/docs/component-splitting/) if you're doing server side rendering. Let's see how we can load them lazily
+
+  ```jsx
+  // With React.Lazy
+  import { Suspense } from 'react';
+
+  const LargeSearchInput = React.lazy(() => import('@razorpay/blade/components'));
+  const SmallSearchInput = React.lazy(() => import('@razorpay/blade/components'));
+
+  const SearchScreen = () => (
+    <Suspense fallback={<div>Loading...</div>}>
+      {isMobile ? <SmallSearchInput /> : <LargeSearchInput />}
```

Comment:
This is a similar question to what Chaitanya has posted above. While it can do that but there are a couple of downsides to it:
1. Every client might have a different way of lazy loading. for csr people might use `React.Lazy` but for SSR they might have to use `loadable` components. Now the moment we generalize this people will have concerns and differences of opinions plus unnecessary dependencies because I can just use `React.lazy` for csr why shall someone install loadable. 
2. Also there are high chances that the props will be completely different because the LargeSearchInput might have auto suggestions as drop down but the small search input might have recent searches which might come from a backend API so if we handle this internally we'll have to create another wrapper which will accept the props for both the variant of the component.

Reply:
finally, [answered here](https://github.com/razorpay/blade/pull/416#issuecomment-1056790157) in detail